core/solver/tfqmr.cpp

```diff
@@ -0,0 +1,248 @@
+    using std::swap;
+    using Vector = matrix::Dense<ValueType>;
+    ASSERT_CONFORMANT(system_matrix_, b);
+    ASSERT_EQUAL_DIMENSIONS(b, x);
+
+    auto exec = this->get_executor();
+
+    auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+    auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+
+    auto dense_b = as<Vector>(b);
+    auto dense_x = as<Vector>(x);
+    auto r = Vector::create_with_config_of(dense_b);
+    auto r0 = Vector::create_with_config_of(dense_b);
+    auto d = Vector::create_with_config_of(dense_b);
+    auto v = Vector::create_with_config_of(dense_b);
+    auto u_m = Vector::create_with_config_of(dense_b);
+    auto u_mp1 = Vector::create_with_config_of(dense_b);
+    auto w = Vector::create_with_config_of(dense_b);
+    auto Ad = Vector::create_with_config_of(dense_b);
+    auto Au = Vector::create_with_config_of(dense_b);
+    auto Au_new = Vector::create_with_config_of(dense_b);
+    auto pu_m = Vector::create_with_config_of(dense_b);
+
+    auto alpha = Vector::create(exec, 1, dense_b->get_num_cols());
+    auto beta = Vector::create_with_config_of(alpha.get());
+    auto sigma = Vector::create_with_config_of(alpha.get());
+    auto rho_old = Vector::create_with_config_of(alpha.get());
+    auto rho = Vector::create_with_config_of(alpha.get());
+    auto taut = Vector::create_with_config_of(alpha.get());
+    auto tau = Vector::create_with_config_of(alpha.get());
+    auto nomw = Vector::create_with_config_of(alpha.get());
+    auto theta = Vector::create_with_config_of(alpha.get());
+    auto eta = Vector::create_with_config_of(alpha.get());
+    auto rov = Vector::create_with_config_of(alpha.get());
+
+    auto master_tau =
+        Vector::create(exec->get_master(), 1, dense_b->get_num_cols());
+    auto starting_tau = Vector::create_with_config_of(master_tau.get());
+
+    // TODO: replace this with automatic merged kernel generator
+    exec->run(TemplatedOperation<ValueType>::make_initialize_operation(
+        dense_b, r.get(), r0.get(), u_m.get(), u_mp1.get(), pu_m.get(),
+        Au.get(), Ad.get(), w.get(), v.get(), d.get(), taut.get(),
+        rho_old.get(), rho.get(), alpha.get(), beta.get(), taut.get(),
+        sigma.get(), rov.get(), eta.get(), nomw.get(), theta.get()));
+    // r = dense_b
+    // r0 = u_m = w = r
+    // Ad = d = 0
+    // theta = eta = rov = alpha = beta = sigma = 1.0
+
+    system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(), r.get());
+    r0->copy_from(r.get());
+    r->compute_dot(r.get(), tau.get());
+    w->copy_from(r.get());
+    u_m->copy_from(r.get());
+    starting_tau->copy_from(tau.get());
+    rho->copy_from(tau.get());
+    rho_old->copy_from(tau.get());
+    taut->copy_from(tau.get());
+    preconditioner_->apply(u_m.get(), pu_m.get());
+    system_matrix_->apply(pu_m.get(), v.get());
+    Au->copy_from(v.get());
+    Ad->copy_from(d.get());
+
+    for (int iter = 0; iter < max_iters_; ++iter) {
+        if (iter % 2 == 0) {
```

Comment:
I tried to keep the code as close as possible to the MATLAB code. The idea was that this could be used by users to match the functionality. I you outlined in #29 there may be another, more efficient way to go. 
> You can merge more kernels than you have currently merged.

Yes, I think we had this discussion on the phone. I would like to see timings to decide whether it is a good idea. But sure, I can do that.
> You have quite a few additional operations that are completely unnecessary.

Not sure about that. For example, I think
> [u_mp1] = u_m - [alpha] * v                                     ! ERROR: the result is never used

is needed in the update step 
>pu_m = M^-1 [u_mp1]                     <- apply

for the even iteration counts.

> you actually need only 5 merged kernels for 2 iterations (instead of 7 for 1)

Yes, we have more merging in MAGMA-sparse. And I wanted to see the performance difference because less merging is easier to understand.

Reply:
> I tried to keep the code as close as possible to the MATLAB code. The idea was that this could be used by users to match the functionality.

Functionality isn't dictated by implementation details, we shouldn't blindly stick to something if it doesn't work well.

> Yes, I think we had this discussion on the phone. I would like to see timings to decide whether it is a good idea. But sure, I can do that.

I am aware of our discussion:

1. Even if you keep scalar updates separate from vector updates, you can still merge more.
2. Either way, you should merge _Ginkgo_ kernels, and have some of them call two _CUDA_ kernels. This performance issue is an implementation detail, which can differ for different devices, and shouldn't leak to the top-level solver implementation.

> Not sure about that. For example, I think
> 
>     [u_mp1] = u_m - [alpha] * v ! ERROR: the result is never used
> 
> is needed in the update step
> 
>     pu_m = M^-1 [u_mp1] <- apply
> 
> for the even iteration counts.

`u_mp1` gets updated before it's used here: `[u_mp1] = w + [beta] * u_m`.  Just search for `u_mp1` in any of the codes (the original file, my transcript of the code, or the unrolled one), and you'll see that there are two consecutive assigns to `u_mp1`.
This is something that's hard to notice in the original version of the code due to the convoluted implementation. If that's based on MATLAB's code, then their implementation is quite horrible.
Please take some more time to look through my comments (at least the error, and opt ones), and point out __exactly__ what's wrong if something is wrong with them.

> less merging is easier to understand.

I disagree, we have comments below, so you don't look at the merged call anyway. Where's the difference for the user if he sees calls to `step_1 ..., step_5` as oposed to `step_1, ...., step_7`?
Also this code has far greater issues when it comes to "understanding".